File: content/manual/rapid/source/includes/aggregation/aggregation-examples/unpack-arrays/full-files/unpack-arrays.c
#include <stdio.h>
#include <bson/bson.h>
#include <mongoc/mongoc.h>

int main(void)
{
    mongoc_init();

    char *uri = "<connection string>";
    mongoc_client_t *client = mongoc_client_new(uri);
    // start-insert-orders
    mongoc_collection_t *orders = mongoc_client_get_collection(client, "agg_tutorials_db", "orders");

    {
        bson_t *filter = bson_new();
        bson_error_t error;
        if (!mongoc_collection_delete_many(orders, filter, NULL, NULL, &error))
        {
            fprintf(stderr, "Delete error: %s\n", error.message);
        }
        bson_destroy(filter);
    }

    {
        size_t num_docs = 4;
        bson_t *docs[num_docs];

        docs[0] = BCON_NEW(
            "order_id", BCON_INT64(6363763262239),
            "products", "[",
            "{",
            "prod_id", BCON_UTF8("abc12345"),
            "name", BCON_UTF8("Asus Laptop"),
            "price", BCON_INT32(431),
            "}",
            "{",
            "prod_id", BCON_UTF8("def45678"),
            "name", BCON_UTF8("Karcher Hose Set"),
            "price", BCON_INT32(22),
            "}",
            "]");

        docs[1] = BCON_NEW(
            "order_id", BCON_INT64(1197372932325),
            "products", "[",
            "{",
            "prod_id", BCON_UTF8("abc12345"),
            "name", BCON_UTF8("Asus Laptop"),
            "price", BCON_INT32(429),
            "}",
            "]");

        docs[2] = BCON_NEW(
            "order_id", BCON_INT64(9812343774839),
            "products", "[",
            "{",
            "prod_id", BCON_UTF8("pqr88223"),
            "name", BCON_UTF8("Morphy Richards Food Mixer"),
            "price", BCON_INT32(431),
            "}",
            "{",
            "prod_id", BCON_UTF8("def45678"),
            "name", BCON_UTF8("Karcher Hose Set"),
            "price", BCON_INT32(21),
            "}",
            "]");

        docs[3] = BCON_NEW(
            "order_id", BCON_INT64(4433997244387),
            "products", "[",
            "{",
            "prod_id", BCON_UTF8("def45678"),
            "name", BCON_UTF8("Karcher Hose Set"),
            "price", BCON_INT32(23),
            "}",
            "{",
            "prod_id", BCON_UTF8("jkl77336"),
            "name", BCON_UTF8("Picky Pencil Sharpener"),
            "price", BCON_INT32(1),
            "}",
            "{",
            "prod_id", BCON_UTF8("xyz11228"),
            "name", BCON_UTF8("Russell Hobbs Chrome Kettle"),
            "price", BCON_INT32(16),
            "}",
            "]");

        bson_error_t error;
        if (!mongoc_collection_insert_many(orders, (const bson_t **)docs, num_docs, NULL, NULL, &error))
        {
            fprintf(stderr, "Insert error: %s\n", error.message);
        }

        for (int i = 0; i < num_docs; i++)
        {
            bson_destroy(docs[i]);
        }
    }
    // end-insert-orders

    {
        const bson_t *doc;
        bson_t *pipeline = BCON_NEW(
            "pipeline", "[",
            // start-unwind
            "{", "$unwind", "{", "path", BCON_UTF8("$products"), "}", "}",
            // end-unwind
            // start-match
            "{", "$match", "{", "products.price", "{", "$gt", BCON_INT32(15), "}", "}", "}",
            // end-match
            // start-group
            "{", "$group", "{",
            "_id", BCON_UTF8("$products.prod_id"),
            "product", "{", "$first", BCON_UTF8("$products.name"), "}",
            "total_value", "{", "$sum", BCON_UTF8("$products.price"), "}",
            "quantity", "{", "$sum", BCON_INT32(1), "}",
            "}", "}",
            // end-group
            // start-set
            "{", "$set", "{", "product_id", BCON_UTF8("$_id"), "}", "}",
            // end-set
            // start-unset
            "{", "$unset", "[", BCON_UTF8("_id"), "]", "}",
            // end-unset
            "]");

        // start-run-agg
        mongoc_cursor_t *results =
            mongoc_collection_aggregate(orders, MONGOC_QUERY_NONE, pipeline, NULL, NULL);

        bson_destroy(pipeline);
        // end-run-agg

        while (mongoc_cursor_next(results, &doc))
        {
            char *str = bson_as_canonical_extended_json(doc, NULL);
            printf("%s\n", str);
            bson_free(str);
        }
        bson_error_t error;
        if (mongoc_cursor_error(results, &error))
        {
            fprintf(stderr, "Aggregation error: %s\n", error.message);
        }

        mongoc_cursor_destroy(results);
    }

    // start-cleanup
    mongoc_collection_destroy(orders);
    // end-cleanup
    mongoc_client_destroy(client);
    mongoc_cleanup();

    return EXIT_SUCCESS;
}

File: content/manual/rapid/source/includes/aggregation/aggregation-examples/unpack-arrays/full-files/unpack-arrays.rs
use mongodb::{
    bson::doc,
    Client,
    Collection,
};
use serde::{Deserialize, Serialize};
use futures::stream::TryStreamExt;

// start-structs
#[derive(Debug, Serialize, Deserialize)]
struct Product {
    prod_id: String,
    name: String,
    price: i32,
}

#[derive(Debug, Serialize, Deserialize)]
struct Order {
    order_id: i64,
    products: Vec<Product>,
}
// end-structs

#[tokio::main]
async fn main() -> mongodb::error::Result<()> {
    let uri = "<connection string>";
    let client = Client::with_uri_str(uri).await?;

    let agg_db = client.database("agg_tutorials_db");
    agg_db.drop().await?;

    // start-insert-orders
    let orders_coll: Collection<Order> = agg_db.collection("orders");
    orders.delete_many(doc! {}).await?;

    let orders = vec![
        Order {
            order_id: 6363763262239,
            products: vec![
                Product {
                    prod_id: "abc12345".to_string(),
                    name: "Asus Laptop".to_string(),
                    price: 431,
                },
                Product {
                    prod_id: "def45678".to_string(),
                    name: "Karcher Hose Set".to_string(),
                    price: 22,
                },
            ],
        },
        Order {
            order_id: 1197372932325,
            products: vec![Product {
                prod_id: "abc12345".to_string(),
                name: "Asus Laptop".to_string(),
                price: 429,
            }],
        },
        Order {
            order_id: 9812343774839,
            products: vec![
                Product {
                    prod_id: "pqr88223".to_string(),
                    name: "Morphy Richards Food Mixer".to_string(),
                    price: 431,
                },
                Product {
                    prod_id: "def45678".to_string(),
                    name: "Karcher Hose Set".to_string(),
                    price: 21,
                },
            ],
        },
        Order {
            order_id: 4433997244387,
            products: vec![
                Product {
                    prod_id: "def45678".to_string(),
                    name: "Karcher Hose Set".to_string(),
                    price: 23,
                },
                Product {
                    prod_id: "jkl77336".to_string(),
                    name: "Picky Pencil Sharpene".to_string(),
                    price: 1,
                },
                Product {
                    prod_id: "xyz11228".to_string(),
                    name: "Russell Hobbs Chrome Kettle".to_string(),
                    price: 16,
                },
            ],
        },
    ];

    orders_coll.insert_many(orders).await?;
    // end-insert-orders

    let mut pipeline = Vec::new();

    // start-unwind
    pipeline.push(doc! {
        "$unwind": {
            "path": "$products"
        }
    });
    // end-unwind

    // start-match
    pipeline.push(doc! {
        "$match": {
            "products.price": { "$gt": 15 }
        }
    });
    // end-match

    // start-group
    pipeline.push(doc! {
        "$group": {
            "_id": "$products.prod_id",
            "product": { "$first": "$products.name" },
            "total_value": { "$sum": "$products.price" },
            "quantity": { "$sum": 1 }
        }
    });
    // end-group

    // start-set
    pipeline.push(doc! {
        "$set": {
            "prod_id": "$_id"
        }
    });
    // end-set

    // start-unset
    pipeline.push(doc! {
        "$unset": ["_id"]
    });
    // end-unset

    // start-run-agg
    let mut cursor = orders_coll.aggregate(pipeline).await?;
    // end-run-agg

    while let Some(result) = cursor.try_next().await? {
        println!("* {:?}", result);
    }

    Ok(())
}
